### PR TITLE
connection & routing subsystems

### PR DESCRIPTION
**Pull request by charlierudolph**

### CODE REVIEW

File: go/exocom/features/register-service.feature [hunk 1]
```diff
@@ -23,13 +23,12 @@ Feature: Manage new instances of services
...
-          "users": {
+        "clients": [
+          {
             "role": "users",
-            "serviceType": "users",
-            "internalNamespace": "foo"
+            "serviceType": "users"
           }
-        },
+        ],
```
Comment (charlierudolph) [inline go/exocom/features/register-service.feature]: Updated these as the key is redundant and the internalNamespace is available in the routes

File: go/exocom/src/logger/logger.go [hunk 2]
```diff
@@ -52,29 +51,14 @@ func (l *Logger) Header(text string) error {
...
-	}
-	for service, route := range routes {
-		err := l.Log("  --[ " + service + " ]-> " + strings.Join(route.Receives, " + "))
-		if err != nil {
-			return err
-		}
-	}
-	return nil
-}
-
```
Comment (charlierudolph) [inline go/exocom/src/logger/logger.go]: This function was unused

File: go/exocom/src/logger/logger_test.go [hunk 3]
```diff
@@ -72,32 +72,18 @@ var _ = Describe("Logger", func() {
...
-			routes := clientRegistry.Routes{
-				"users": usersRoute,
-			}
-			result := GetOutput(func(testLogger *logger.Logger) {
-				err := testLogger.RoutingSetup(routes)
-				Expect(err).To(BeNil())
-			})
-			Expect(result).To(Equal("Receiving Routing Setup\n  --[ users ]-> users.create\n"))
-		})
-	})
```
Comment (charlierudolph) [inline go/exocom/src/logger/logger_test.go]: Removing the test for the unused function

File: go/exocom/features/register-service.feature (hunk 1 above)
Comment (kevgo) [inline go/exocom/features/register-service.feature]: Yes, these can be updated if you see a way to do this better.

File: go/exocom/src/connection/service.go [hunk 4]
```diff
@@ -0,0 +1,73 @@
...
+}
+
+// BootstrapService creates a new Service which will alert the manager when the
+// service registers (by sending its first message) and deregisters (by disconnecting)
+func BootstrapService(manager *Manager, socket *websocket.Conn) {
+	service := Service{
+		manager: manager,
+		socket:  socket,
+	}
+	fmt.Println("bootstraping service")
```
Comment (kevgo) [inline go/exocom/src/connection/service.go]: remove debug statement?

File: go/exocom/src/connection/service.go [hunk 5]
```diff
@@ -0,0 +1,73 @@
...
+	return s.socket.WriteMessage(websocket.TextMessage, serializedBytes)
+}
+
+// Helpers
+
+func (s *Service) listen() {
+	utils.ListenForMessages(s.socket, func(message structs.Message) error {
+		fmt.Println("received message", message)
+		if message.Name == "exocom.register-service" {
+			var err error
```
Comment (kevgo) [inline go/exocom/src/connection/service.go]: extract this into a `handleRegisterMessage` method?

File: go/exocom/src/connection/service.go [hunk 6]
```diff
@@ -0,0 +1,73 @@
...
+				fmt.Println("registering service")
+				s.manager.registerService(s)
+			} else {
+				s.manager.logError(err)
+			}
+		} else {
+			s.manager.onMessage(message)
+		}
+		return nil
+	}, func(err error) {
```
Comment (kevgo) [inline go/exocom/src/connection/service.go]: Is this callback really the best way to go about this? Can we use channels to signal errors?

Comment (charlierudolph) [inline go/exocom/src/connection/service.go]: Sounds good. This is an existing helper that is used in a few other places. I will update it to use channels in a separate PR

Comment (charlierudolph) [inline go/exocom/src/connection/service.go]: extracted to #199

File: go/exocom/src/connection/service.go (hunk 4 above)
Comment (charlierudolph) [inline go/exocom/src/connection/service.go]: done

File: go/exocom/src/connection/service.go (hunk 5 above)
Comment (charlierudolph) [inline go/exocom/src/connection/service.go]: done

**Comment by charlierudolph** [comment]
@kevgo all comments addressed

**Pull request merged**